src/common/serializer/buffered_file_writer.cpp

```diff
@@ -22,16 +23,32 @@ idx_t BufferedFileWriter::GetTotalWritten() {
 }
 
 void BufferedFileWriter::WriteData(const_data_ptr_t buffer, idx_t write_size) {
-	// first copy anything we can from the buffer
-	const_data_ptr_t end_ptr = buffer + write_size;
-	while (buffer < end_ptr) {
-		idx_t to_write = MinValue<idx_t>((end_ptr - buffer), FILE_BUFFER_SIZE - offset);
-		D_ASSERT(to_write > 0);
-		memcpy(data.get() + offset, buffer, to_write);
-		offset += to_write;
-		buffer += to_write;
-		if (offset == FILE_BUFFER_SIZE) {
-			Flush();
+	if (write_size >= (FILE_BUFFER_SIZE + offset)) {
```

Comment:
this should be `- offset` no?
We're trying to assess whether this write would be bigger than the remaining space we have left inside our buffer?

Reply:
No it's `+offset` here because otherwise if the size is 4096 and offset is 1 it will:
1. Copy 4095 bytes into the internal buffer
2. Flush (call the underlying filesystem)
3. Write in the underlying filesystem the remaining byte

In this case I preferred to go to the else case and keep the remaining in the, not yet flush, internal buffer.

Reply:
Shouldn't it be `2*FILE_BUFFER_SIZE - offset`?
We want to take the direct io path only if we have more data that needed to:
- fill the remainder of the buffer (i.e. FILE_BUFFER_SIZE - offset)
- fill a complete buffer after this (i.e. FILE_BUFFER_SIZE)

with `FILE_BUFFER_SIZE + offset` consider a write of `FILE_BUFFER_SIZE + 2` where the offset is 1. 
We now take the direct io path, but we need to copy most in the buffer anyway. Then we follow up with a super small write, no?

Reply:
Hi @samansmink !
Yes your are absolutely right !